integrations/tensorflow/iree_tf_compiler/TFL/RetainCallOnceFuncs.cpp

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/Utils/ConversionUtils.h"
+#include "iree_tf_compiler/TFL/Passes.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "tensorflow/compiler/mlir/lite/ir/tfl_ops.h"
+
+namespace mlir {
+namespace iree_integrations {
+namespace TFL {
+namespace {
+
+class RetainCallOnceFuncsPass
+    : public PassWrapper<RetainCallOnceFuncsPass, OperationPass<ModuleOp>> {
+ public:
+  void getDependentDialects(DialectRegistry& registry) const override {
+    registry.insert<mlir::TFL::TensorFlowLiteDialect>();
+  }
+
+  StringRef getArgument() const override {
+    return "iree-tflite-retain-call-once-funcs";
+  }
+
+  StringRef getDescription() const override {
+    return "Guarantees that functions used by tfl.call_once are retained";
+  }
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+
+    llvm::DenseMap<StringRef, FuncOp> funcMap;
+    for (auto func : moduleOp.getOps<mlir::FuncOp>()) {
+      funcMap[func.sym_name()] = func;
+    }
+
+    for (auto func : moduleOp.getOps<mlir::FuncOp>()) {
+      for (auto callOnce : func.getOps<mlir::TFL::CallOnceOp>()) {
+        auto callFunc = funcMap[callOnce.session_init_function()];
```

Comment:
Huh, why isn't this a SymbolRefAttr in the first place? I think it's also not clear that attaching these unknown attributes to an operation and expecting passes to preserve them can be load bearing for correctness. See extensive discussion in https://llvm.discourse.group/t/on-querying-an-operations-intrinsic-core-vs-external-user-defined-attributes/4498/

Reply:
I really don't have a better suggestion :-/ Maybe inquire with TFLite folks why this isn't a symbol ref attr? More wanted to make you aware that this is on a shaky foundation